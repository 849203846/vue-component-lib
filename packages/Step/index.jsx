import sfc from '../utils/sfc';

const { def, bem } = sfc('step');

export default def({
  props: {
    title: String,
    description: String,
  },

  data() {
    return {
      index: 0,
    };
  },

  beforeCreate() {
    this.$parent.steps.push(this);
  },

  beforeDestroy() {
    const { steps } = this.$parent;
    const index = steps.indexOf(this);
    if (index >= 0) {
      steps.splice(index, 1);
    }
  },

  render() {
    const { direction } = this.$parent;
    const title = (t, slot) => <div class={bem('title')}>{t || slot()}</div>;
    const description = (d, slot) => <div class={bem('description')}>{d || slot()}</div>;

    return (
      <div class={bem({ [direction]: true })}>
        <div class={bem('head')}>
          <span class={bem('head-inner')}>{this.index}</span>
        </div>
        <div class={bem('main')}>
          {(this.title || this.$scopedSlots.title) && title(this.title, this.$scopedSlots.title)}
          {(this.description || this.$scopedSlots.default) && description(this.description, this.$scopedSlots.default)}
        </div>
      </div>
    );
  },
});
