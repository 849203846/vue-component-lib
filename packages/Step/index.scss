@import "../style/global";

@mixin step-horizontal() {
  z-index: 2;
  padding: 0 $--step-padding-horizontal;

  &:first-of-type {
    padding-left: 0;
  }

  &:last-of-type {
    padding-right: 0;
  }

  @include e(main) {
    margin-left: $--step-main-margin-horizontal;
  }

  @include e(description) {
    margin-top: 4px;
  }
}

@mixin step-vertical() {
  &:last-of-type {
    z-index: 2;
    padding-top: $--step-padding-vertical;
  }

  &:first-of-type {
    @include e(head) {
      padding-bottom: $--step-padding-vertical;
    }
  }

  &:not(:first-of-type):not(:last-of-type) {
    @include e(head) {
      padding: $--step-padding-vertical 0;
    }
    @include e(main) {
      padding-top: $--step-padding-vertical;
    }
  }

  @include e(head) {
    z-index: 2;
    background: inherit;
  }

  @include e(main) {
    margin-left: $--step-main-margin-vertical;
  }

  @include e(description) {
    margin-top: 10px;
  }
}

@include b(step) {
  display: flex;
  align-items: flex-start;
  text-align: left;
  background: inherit;

  @include m(horizontal) {
    @include step-horizontal;
  }

  @include m(vertical) {
    @include step-vertical;
  }

  @include m(auto) {
    @media only screen and (max-width: 767px) {
      @include step-vertical;
    }

    @media only screen and (min-width: 768px) {
      @include step-horizontal;
    }
  }

  @include e(head) {
    flex: none;
  }

  @include e(head-inner) {
    display: inline-flex;
    align-items: center;
    justify-content: center;
    width: $--step-head-size;
    height: $--step-head-size;
    font-size: 14px;
    line-height: 1;
    color: inherit;
    text-align: center;
    user-select: none;
    background-color: #8682f9;
    border-radius: 50%;
  }

  @include e(title) {
    font-size: 16px;
    line-height: $--step-head-size;
  }

  @include e(description) {
    font-size: 12px;
    line-height: 18px;
  }
}
