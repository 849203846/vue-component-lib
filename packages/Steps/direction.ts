enum StepsDirection {
  auto = 'auto',
  horizontal = 'horizontal',
  vertical = 'vertical',
}

export default StepsDirection;
