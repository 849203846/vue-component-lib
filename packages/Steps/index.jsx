import sfc from '../utils/sfc';
import DirectionOptions from './direction';

const { def, bem } = sfc('steps');

export default def({
  props: {
    direction: {
      type: String,
      default: DirectionOptions.auto,
      validator(value) {
        return !!DirectionOptions[value];
      },
    },
  },

  data() {
    return {
      steps: [],
    };
  },

  computed: {
    classes() {
      return bem({
        [this.direction]: true,
        'only-one': this.steps.length <= 1,
      });
    },
  },

  watch: {
    steps(to) {
      to.forEach((item, index) => {
        item.index = index + 1;
      });
    },
  },

  render() {
    return (
      <div class={this.classes}>
        {this.$scopedSlots.default && this.$scopedSlots.default()}
      </div>
    );
  },
});
