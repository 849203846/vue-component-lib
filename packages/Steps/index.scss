@import "../style/global";

@include b(steps) {
  $root: &;

  position: relative;
  display: flex;
  justify-content: space-between;
  background-color: inherit;

  &::before {
    position: absolute;
    content: "";
    background: $--step-line-color;
  }

  @include m(only-one) {
    &::before {
      display: none;
    }
  }

  @include m(horizontal) {
    flex-direction: row;

    &#{$root}::before {
      top: $--step-head-size / 2;
      right: 0;
      left: 0;
      height: $--step-line-size;
      transform: translateY(-50%);
    }
  }

  @include m(vertical) {
    flex-direction: column;

    &#{$root}::before {
      top: 0;
      bottom: 0;
      left: $--step-head-size / 2;
      width: $--step-line-size;
      transform: translateX(-50%);
    }
  }

  @include m(auto) {
    flex-direction: column;

    @media only screen and (max-width: 767px) {
      &#{$root}::before {
        top: 0;
        bottom: 0;
        left: $--step-head-size / 2;
        width: $--step-line-size;
        transform: translateX(-50%);
      }
    }

    @media only screen and (min-width: 768px) {
      flex-direction: row;

      &#{$root}::before {
        top: $--step-head-size / 2;
        right: 0;
        left: 0;
        height: $--step-line-size;
        transform: translateY(-50%);
      }
    }
  }
}
