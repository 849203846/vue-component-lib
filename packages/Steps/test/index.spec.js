import { mount } from '@vue/test-utils';
import DirectionOptions from '../direction';
import Steps from '..';
import Step from '../../Step';

describe('Component Steps', () => {
  it('default direction and empty slot', () => {
    const wrapper = mount(Steps);
    expect(wrapper.classes()).toContain('v-steps');
    expect(wrapper.classes()).toContain('v-steps--auto');
    expect(wrapper.classes()).toContain('v-steps--only-one');
    expect(wrapper.isEmpty()).toBeTruthy();
    expect(wrapper.element).toMatchSnapshot();
  });

  it('correct direction', () => {
    const wrapper = mount(Steps, {
      propsData: {
        direction: DirectionOptions.horizontal,
      },
    });
    expect(wrapper.classes()).toContain('v-steps');
    expect(wrapper.classes()).toContain(`v-steps--${DirectionOptions.horizontal}`);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('correct slot', () => {
    const wrapper = mount({
      components: {
        Steps,
        Step,
      },
      template: `
        <steps>
          <step title="1" description="description 1"></step>
          <step>
            <template v-slot:title>test</template>
            slot description test
          </step>
          <step></step>
        </steps>
      `,
    });
    expect(wrapper.element).toMatchSnapshot();

    const stepsIns = wrapper.find(Steps);
    expect(stepsIns.vm.steps.length).toBe(3);
    stepsIns.destroy();
    expect(stepsIns.vm.steps.length).toBe(0);
  });

  it('update step', () => {
    const wrapper = mount(Steps);
    expect(wrapper.vm.steps.length).toBe(0);

    const genItem = () => ({ index: 0 });

    wrapper.vm.steps.push(genItem());
    expect(wrapper.vm.steps.length).toBe(1);
    expect(wrapper.vm.steps[0].index).toBe(1);
    expect(wrapper.classes()).toContain('v-steps--only-one');

    wrapper.vm.steps.push(genItem());
    expect(wrapper.vm.steps.length).toBe(2);
    expect(wrapper.vm.steps[0].index).toBe(1);
    expect(wrapper.vm.steps[1].index).toBe(2);
    expect(wrapper.classes().includes('v-steps--only-one')).toBeFalsy();
  });
});
