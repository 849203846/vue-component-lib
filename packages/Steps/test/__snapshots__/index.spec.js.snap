// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Steps correct direction 1`] = `
<div
  class="v-steps v-steps--horizontal v-steps--only-one"
/>
`;

exports[`Component Steps correct slot 1`] = `
<div
  class="v-steps v-steps--auto v-steps--only-one"
>
  <div
    class="v-step v-step--auto"
  >
    <div
      class="v-step__head"
    >
      <span
        class="v-step__head-inner"
      >
        1
      </span>
    </div>
    <div
      class="v-step__main"
    >
      <div
        class="v-step__title"
      >
        1
      </div>
      <div
        class="v-step__description"
      >
        description 1
      </div>
    </div>
  </div>
   
  <div
    class="v-step v-step--auto"
  >
    <div
      class="v-step__head"
    >
      <span
        class="v-step__head-inner"
      >
        2
      </span>
    </div>
    <div
      class="v-step__main"
    >
      <div
        class="v-step__title"
      >
        test
      </div>
      <div
        class="v-step__description"
      >
        
            slot description test
          
      </div>
    </div>
  </div>
   
  <div
    class="v-step v-step--auto"
  >
    <div
      class="v-step__head"
    >
      <span
        class="v-step__head-inner"
      >
        0
      </span>
    </div>
    <div
      class="v-step__main"
    />
  </div>
</div>
`;

exports[`Component Steps default direction and empty slot 1`] = `
<div
  class="v-steps v-steps--auto v-steps--only-one"
/>
`;
